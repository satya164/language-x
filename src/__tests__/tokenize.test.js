/* @flow */

const dedent = require('dedent');
const tokenize = require('../tokenize');

it('tokenizes type declaration', () => {
  expect(
    tokenize(dedent`
    type Bar = Boolean | String | Number
    type Maybe Number = Nothing | Number
    type Users = List User
  `)
  ).toMatchSnapshot();
});

it('tokenizes let declaration', () => {
  expect(
    tokenize(dedent`
    let a = 10
  `)
  ).toMatchSnapshot();
});

it('tokenizes various types of numbers', () => {
  expect(
    tokenize(dedent`
    let a = 10
    let a = 10.30
    let a = +10
    let a = -10
    let a = .30
    let a = +.30
    let a = -.30
    let a = +0.30
    let a = -0.30
  `)
  ).toMatchSnapshot();
});

it('tokenizes strings', () => {
  expect(
    tokenize(dedent`
    let a = "Hello world"
    let a = "Numbers in strings +0.30"
    let a = "Operators in string (= | + - /)"
    let a = "Keywords such as let in string"
  `)
  ).toMatchSnapshot();
});
