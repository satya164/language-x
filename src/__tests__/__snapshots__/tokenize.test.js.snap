// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenizes let declaration 1`] = `
Array [
  Object {
    "loc": Object {
      "column": 1,
      "line": 0,
    },
    "type": "keyword",
    "value": "let",
  },
  Object {
    "loc": Object {
      "column": 4,
      "line": 0,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 5,
      "line": 0,
    },
    "type": "identifier",
    "value": "a",
  },
  Object {
    "loc": Object {
      "column": 6,
      "line": 0,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 7,
      "line": 0,
    },
    "type": "operator",
    "value": "=",
  },
  Object {
    "loc": Object {
      "column": 8,
      "line": 0,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 9,
      "line": 0,
    },
    "type": "number",
    "value": "10",
  },
]
`;

exports[`tokenizes strings 1`] = `
Array [
  Object {
    "loc": Object {
      "column": 1,
      "line": 0,
    },
    "type": "keyword",
    "value": "let",
  },
  Object {
    "loc": Object {
      "column": 4,
      "line": 0,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 5,
      "line": 0,
    },
    "type": "identifier",
    "value": "a",
  },
  Object {
    "loc": Object {
      "column": 6,
      "line": 0,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 7,
      "line": 0,
    },
    "type": "operator",
    "value": "=",
  },
  Object {
    "loc": Object {
      "column": 8,
      "line": 0,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 10,
      "line": 0,
    },
    "type": "string",
    "value": "Hello world",
  },
  Object {
    "loc": Object {
      "column": 0,
      "line": 1,
    },
    "type": "newline",
    "value": "
",
  },
  Object {
    "loc": Object {
      "column": 1,
      "line": 1,
    },
    "type": "keyword",
    "value": "let",
  },
  Object {
    "loc": Object {
      "column": 4,
      "line": 1,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 5,
      "line": 1,
    },
    "type": "identifier",
    "value": "a",
  },
  Object {
    "loc": Object {
      "column": 6,
      "line": 1,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 7,
      "line": 1,
    },
    "type": "operator",
    "value": "=",
  },
  Object {
    "loc": Object {
      "column": 8,
      "line": 1,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 10,
      "line": 1,
    },
    "type": "string",
    "value": "Numbers in strings +0.30",
  },
  Object {
    "loc": Object {
      "column": 0,
      "line": 2,
    },
    "type": "newline",
    "value": "
",
  },
  Object {
    "loc": Object {
      "column": 1,
      "line": 2,
    },
    "type": "keyword",
    "value": "let",
  },
  Object {
    "loc": Object {
      "column": 4,
      "line": 2,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 5,
      "line": 2,
    },
    "type": "identifier",
    "value": "a",
  },
  Object {
    "loc": Object {
      "column": 6,
      "line": 2,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 7,
      "line": 2,
    },
    "type": "operator",
    "value": "=",
  },
  Object {
    "loc": Object {
      "column": 8,
      "line": 2,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 10,
      "line": 2,
    },
    "type": "string",
    "value": "Operators in string (= | + - /)",
  },
  Object {
    "loc": Object {
      "column": 0,
      "line": 3,
    },
    "type": "newline",
    "value": "
",
  },
  Object {
    "loc": Object {
      "column": 1,
      "line": 3,
    },
    "type": "keyword",
    "value": "let",
  },
  Object {
    "loc": Object {
      "column": 4,
      "line": 3,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 5,
      "line": 3,
    },
    "type": "identifier",
    "value": "a",
  },
  Object {
    "loc": Object {
      "column": 6,
      "line": 3,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 7,
      "line": 3,
    },
    "type": "operator",
    "value": "=",
  },
  Object {
    "loc": Object {
      "column": 8,
      "line": 3,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 10,
      "line": 3,
    },
    "type": "string",
    "value": "Keywords such as let in string",
  },
]
`;

exports[`tokenizes type declaration 1`] = `
Array [
  Object {
    "loc": Object {
      "column": 1,
      "line": 0,
    },
    "type": "keyword",
    "value": "type",
  },
  Object {
    "loc": Object {
      "column": 5,
      "line": 0,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 6,
      "line": 0,
    },
    "type": "identifier",
    "value": "Foo",
  },
  Object {
    "loc": Object {
      "column": 0,
      "line": 1,
    },
    "type": "newline",
    "value": "
",
  },
  Object {
    "loc": Object {
      "column": 1,
      "line": 1,
    },
    "type": "keyword",
    "value": "type",
  },
  Object {
    "loc": Object {
      "column": 5,
      "line": 1,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 6,
      "line": 1,
    },
    "type": "identifier",
    "value": "Bar",
  },
  Object {
    "loc": Object {
      "column": 9,
      "line": 1,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 10,
      "line": 1,
    },
    "type": "operator",
    "value": "=",
  },
  Object {
    "loc": Object {
      "column": 11,
      "line": 1,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 12,
      "line": 1,
    },
    "type": "identifier",
    "value": "Boolean",
  },
  Object {
    "loc": Object {
      "column": 19,
      "line": 1,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 20,
      "line": 1,
    },
    "type": "operator",
    "value": "|",
  },
  Object {
    "loc": Object {
      "column": 21,
      "line": 1,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 22,
      "line": 1,
    },
    "type": "identifier",
    "value": "String",
  },
  Object {
    "loc": Object {
      "column": 28,
      "line": 1,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 29,
      "line": 1,
    },
    "type": "operator",
    "value": "|",
  },
  Object {
    "loc": Object {
      "column": 30,
      "line": 1,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 31,
      "line": 1,
    },
    "type": "identifier",
    "value": "Number",
  },
  Object {
    "loc": Object {
      "column": 0,
      "line": 2,
    },
    "type": "newline",
    "value": "
",
  },
  Object {
    "loc": Object {
      "column": 1,
      "line": 2,
    },
    "type": "keyword",
    "value": "type",
  },
  Object {
    "loc": Object {
      "column": 5,
      "line": 2,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 6,
      "line": 2,
    },
    "type": "identifier",
    "value": "Maybe",
  },
  Object {
    "loc": Object {
      "column": 11,
      "line": 2,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 12,
      "line": 2,
    },
    "type": "identifier",
    "value": "Number",
  },
  Object {
    "loc": Object {
      "column": 18,
      "line": 2,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 19,
      "line": 2,
    },
    "type": "operator",
    "value": "=",
  },
  Object {
    "loc": Object {
      "column": 20,
      "line": 2,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 21,
      "line": 2,
    },
    "type": "identifier",
    "value": "Nothing",
  },
  Object {
    "loc": Object {
      "column": 28,
      "line": 2,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 29,
      "line": 2,
    },
    "type": "operator",
    "value": "|",
  },
  Object {
    "loc": Object {
      "column": 30,
      "line": 2,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 31,
      "line": 2,
    },
    "type": "identifier",
    "value": "Number",
  },
  Object {
    "loc": Object {
      "column": 0,
      "line": 3,
    },
    "type": "newline",
    "value": "
",
  },
  Object {
    "loc": Object {
      "column": 1,
      "line": 3,
    },
    "type": "keyword",
    "value": "type",
  },
  Object {
    "loc": Object {
      "column": 5,
      "line": 3,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 6,
      "line": 3,
    },
    "type": "identifier",
    "value": "Users",
  },
  Object {
    "loc": Object {
      "column": 11,
      "line": 3,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 12,
      "line": 3,
    },
    "type": "operator",
    "value": "=",
  },
  Object {
    "loc": Object {
      "column": 13,
      "line": 3,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 14,
      "line": 3,
    },
    "type": "identifier",
    "value": "List",
  },
  Object {
    "loc": Object {
      "column": 18,
      "line": 3,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 19,
      "line": 3,
    },
    "type": "identifier",
    "value": "User",
  },
]
`;

exports[`tokenizes various types of numbers 1`] = `
Array [
  Object {
    "loc": Object {
      "column": 1,
      "line": 0,
    },
    "type": "keyword",
    "value": "let",
  },
  Object {
    "loc": Object {
      "column": 4,
      "line": 0,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 5,
      "line": 0,
    },
    "type": "identifier",
    "value": "a",
  },
  Object {
    "loc": Object {
      "column": 6,
      "line": 0,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 7,
      "line": 0,
    },
    "type": "operator",
    "value": "=",
  },
  Object {
    "loc": Object {
      "column": 8,
      "line": 0,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 9,
      "line": 0,
    },
    "type": "number",
    "value": "10",
  },
  Object {
    "loc": Object {
      "column": 0,
      "line": 1,
    },
    "type": "newline",
    "value": "
",
  },
  Object {
    "loc": Object {
      "column": 1,
      "line": 1,
    },
    "type": "keyword",
    "value": "let",
  },
  Object {
    "loc": Object {
      "column": 4,
      "line": 1,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 5,
      "line": 1,
    },
    "type": "identifier",
    "value": "a",
  },
  Object {
    "loc": Object {
      "column": 6,
      "line": 1,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 7,
      "line": 1,
    },
    "type": "operator",
    "value": "=",
  },
  Object {
    "loc": Object {
      "column": 8,
      "line": 1,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 9,
      "line": 1,
    },
    "type": "number",
    "value": "10.30",
  },
  Object {
    "loc": Object {
      "column": 0,
      "line": 2,
    },
    "type": "newline",
    "value": "
",
  },
  Object {
    "loc": Object {
      "column": 1,
      "line": 2,
    },
    "type": "keyword",
    "value": "let",
  },
  Object {
    "loc": Object {
      "column": 4,
      "line": 2,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 5,
      "line": 2,
    },
    "type": "identifier",
    "value": "a",
  },
  Object {
    "loc": Object {
      "column": 6,
      "line": 2,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 7,
      "line": 2,
    },
    "type": "operator",
    "value": "=",
  },
  Object {
    "loc": Object {
      "column": 8,
      "line": 2,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 9,
      "line": 2,
    },
    "type": "number",
    "value": "+10",
  },
  Object {
    "loc": Object {
      "column": 0,
      "line": 3,
    },
    "type": "newline",
    "value": "
",
  },
  Object {
    "loc": Object {
      "column": 1,
      "line": 3,
    },
    "type": "keyword",
    "value": "let",
  },
  Object {
    "loc": Object {
      "column": 4,
      "line": 3,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 5,
      "line": 3,
    },
    "type": "identifier",
    "value": "a",
  },
  Object {
    "loc": Object {
      "column": 6,
      "line": 3,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 7,
      "line": 3,
    },
    "type": "operator",
    "value": "=",
  },
  Object {
    "loc": Object {
      "column": 8,
      "line": 3,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 9,
      "line": 3,
    },
    "type": "number",
    "value": "-10",
  },
  Object {
    "loc": Object {
      "column": 0,
      "line": 4,
    },
    "type": "newline",
    "value": "
",
  },
  Object {
    "loc": Object {
      "column": 1,
      "line": 4,
    },
    "type": "keyword",
    "value": "let",
  },
  Object {
    "loc": Object {
      "column": 4,
      "line": 4,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 5,
      "line": 4,
    },
    "type": "identifier",
    "value": "a",
  },
  Object {
    "loc": Object {
      "column": 6,
      "line": 4,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 7,
      "line": 4,
    },
    "type": "operator",
    "value": "=",
  },
  Object {
    "loc": Object {
      "column": 8,
      "line": 4,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 9,
      "line": 4,
    },
    "type": "number",
    "value": ".30",
  },
  Object {
    "loc": Object {
      "column": 0,
      "line": 5,
    },
    "type": "newline",
    "value": "
",
  },
  Object {
    "loc": Object {
      "column": 1,
      "line": 5,
    },
    "type": "keyword",
    "value": "let",
  },
  Object {
    "loc": Object {
      "column": 4,
      "line": 5,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 5,
      "line": 5,
    },
    "type": "identifier",
    "value": "a",
  },
  Object {
    "loc": Object {
      "column": 6,
      "line": 5,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 7,
      "line": 5,
    },
    "type": "operator",
    "value": "=",
  },
  Object {
    "loc": Object {
      "column": 8,
      "line": 5,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 9,
      "line": 5,
    },
    "type": "number",
    "value": "+.30",
  },
  Object {
    "loc": Object {
      "column": 0,
      "line": 6,
    },
    "type": "newline",
    "value": "
",
  },
  Object {
    "loc": Object {
      "column": 1,
      "line": 6,
    },
    "type": "keyword",
    "value": "let",
  },
  Object {
    "loc": Object {
      "column": 4,
      "line": 6,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 5,
      "line": 6,
    },
    "type": "identifier",
    "value": "a",
  },
  Object {
    "loc": Object {
      "column": 6,
      "line": 6,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 7,
      "line": 6,
    },
    "type": "operator",
    "value": "=",
  },
  Object {
    "loc": Object {
      "column": 8,
      "line": 6,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 9,
      "line": 6,
    },
    "type": "number",
    "value": "-.30",
  },
  Object {
    "loc": Object {
      "column": 0,
      "line": 7,
    },
    "type": "newline",
    "value": "
",
  },
  Object {
    "loc": Object {
      "column": 1,
      "line": 7,
    },
    "type": "keyword",
    "value": "let",
  },
  Object {
    "loc": Object {
      "column": 4,
      "line": 7,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 5,
      "line": 7,
    },
    "type": "identifier",
    "value": "a",
  },
  Object {
    "loc": Object {
      "column": 6,
      "line": 7,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 7,
      "line": 7,
    },
    "type": "operator",
    "value": "=",
  },
  Object {
    "loc": Object {
      "column": 8,
      "line": 7,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 9,
      "line": 7,
    },
    "type": "number",
    "value": "+0.30",
  },
  Object {
    "loc": Object {
      "column": 0,
      "line": 8,
    },
    "type": "newline",
    "value": "
",
  },
  Object {
    "loc": Object {
      "column": 1,
      "line": 8,
    },
    "type": "keyword",
    "value": "let",
  },
  Object {
    "loc": Object {
      "column": 4,
      "line": 8,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 5,
      "line": 8,
    },
    "type": "identifier",
    "value": "a",
  },
  Object {
    "loc": Object {
      "column": 6,
      "line": 8,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 7,
      "line": 8,
    },
    "type": "operator",
    "value": "=",
  },
  Object {
    "loc": Object {
      "column": 8,
      "line": 8,
    },
    "type": "whitespace",
    "value": " ",
  },
  Object {
    "loc": Object {
      "column": 9,
      "line": 8,
    },
    "type": "number",
    "value": "-0.30",
  },
]
`;
